jetty-http2/http2-common/src/main/java/org/eclipse/jetty/http2/HTTP2Connection.java

```diff
@@ -234,95 +233,108 @@ public Runnable produce()
+                    int filled = fill(getEndPoint(), networkBuffer.getBuffer());
                     if (LOG.isDebugEnabled())
-                        LOG.debug("Dequeued new task {}", task);
-                    if (task != null)
+                        LOG.debug("Filled {} bytes in {}", filled, networkBuffer);
+
+                    if (filled > 0)
+                    {
+                        bytesIn.addAndGet(filled);
+                        parse = true;
+                    }
+                    else if (filled == 0)
                     {
-                        if (released)
-                            releaseNetworkBuffer();
-                        else
-                            networkBuffer = null;
-                        return task;
+                        interested = true;
+                        return null;
                     }
                     else
                     {
-                        if (!released)
-                            networkBuffer = acquireNetworkBuffer();
+                        shutdown = true;
+                        session.onShutdown();
+                        return null;
                     }
                 }
+            }
+            finally
+            {
+                releaseNetworkBuffer();
+                if (interested)
+                    getEndPoint().fillInterested(fillableCallback);
+            }
+        }
 
-                // Here we know that this.buffer is not retained:
-                // either it has been released, or it's a new one.
 
-                int filled = fill(getEndPoint(), networkBuffer.buffer);
+        private void acquireNetworkBuffer()
+        {
+            if (networkBuffer == null)
+            {
+                networkBuffer = new NetworkBuffer();
                 if (LOG.isDebugEnabled())
-                    LOG.debug("Filled {} bytes in {}", filled, networkBuffer);
-
-                if (filled > 0)
-                {
-                    bytesIn.addAndGet(filled);
-                    parse = true;
-                }
-                else if (filled == 0)
-                {
-                    releaseNetworkBuffer();
-                    getEndPoint().fillInterested(fillableCallback);
-                    return null;
-                }
-                else
-                {
-                    releaseNetworkBuffer();
-                    shutdown = true;
-                    session.onShutdown();
-                    return null;
-                }
+                    LOG.debug("Acquired {}", networkBuffer);
             }
         }
 
-        private NetworkBuffer acquireNetworkBuffer()
+        private void reacquireNetworkBuffer()
         {
-            NetworkBuffer networkBuffer = new NetworkBuffer();
+            if (networkBuffer == null)
+                throw new IllegalStateException();
+
+            if (networkBuffer.getBuffer().hasRemaining())
+                throw new IllegalStateException();
```

Comment:
It can happen that `networkBuffer` has remaining bytes.

Reply:
I don't think it can have remaining bytes here.  We only reacquire if we have consumed the current buffer and it is still referred to.   It can happen in release and I have fixed that.